{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f0f3cb9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install -q openai langchain playwright beautifulsoup4 playwright install chroma chromadb tiktoken langchainhub crewai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a30768e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " ········\n"
     ]
    }
   ],
   "source": [
    "from getpass import getpass\n",
    "import os\n",
    "os.environ['OPENAI_API_KEY'] = getpass()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e34a7025-844b-410d-84dd-22bab2fc457f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract text from CDC website\n",
    "from langchain_community.document_loaders.recursive_url_loader import RecursiveUrlLoader\n",
    "from bs4 import BeautifulSoup as Soup\n",
    "\n",
    "import bs4\n",
    "from langchain import hub\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "from langchain.document_loaders import WebBaseLoader\n",
    "from langchain_openai import OpenAIEmbeddings\n",
    "from langchain.schema import StrOutputParser\n",
    "from langchain.schema.runnable import RunnablePassthrough\n",
    "from openai import OpenAI\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from langchain.vectorstores import Chroma\n",
    "import time\n",
    "\n",
    "#Send to GPT4 for cleanup\n",
    "def clean_text_with_gpt4(text):\n",
    "    \"\"\"\n",
    "    This function takes a string of text and uses GPT-4 to clean it up using the OpenAI ChatCompletion API.\n",
    "    It handles large texts by breaking them into smaller chunks.\n",
    "    :param text: String containing the text to be cleaned.\n",
    "    :return: Cleaned text as a string.\n",
    "    \"\"\"\n",
    "    cleaned_texts = []\n",
    "    text_splitter = RecursiveCharacterTextSplitter(chunk_size=2000, chunk_overlap=0)\n",
    "    chunks = text_splitter.split_documents(text)\n",
    "\n",
    "    for chunk in chunks:\n",
    "        try:\n",
    "            response = client.chat.completions.create(\n",
    "                model=\"gpt-4-0314\",  # Assuming using the latest GPT-4 model\n",
    "                messages=[{\"role\": \"system\", \"content\": \"You are a helpful assistant.\"},\n",
    "                          {\"role\": \"user\", \"content\": f\"Please clean up the following text:\\n\\n{chunk}\"}]\n",
    "            )\n",
    "            #print(response)\n",
    "            cleaned_texts.append(response.choices[0].message.content.strip())\n",
    "            #print(cleaned_texts)\n",
    "            time.sleep(1)  # Delay to respect rate limits\n",
    "        except Exception as e:\n",
    "            print(f\"An error occurred: {e}\")\n",
    "            return None\n",
    "    return ' '.join(cleaned_texts)\n",
    "\n",
    "file_name = \"cleaned_texts.txt\"\n",
    "if os.path.exists(file_name):\n",
    "    print(f\"The file {file_name} already exists.\")\n",
    "    try:\n",
    "        with open(file_name, 'r', encoding='utf-8') as file:\n",
    "            content = file.read()\n",
    "            #print(\"File content:\\n\")\n",
    "            #print(content)\n",
    "    except FileNotFoundError:\n",
    "        print(f\"The file {file_name} was not found.\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "else:\n",
    "    url = \"https://www.cdc.gov\"\n",
    "    loader = RecursiveUrlLoader(\n",
    "        url=url, max_depth=2, extractor=lambda x: Soup(x, \"html.parser\").text\n",
    "    )\n",
    "    docs = loader.load()\n",
    "    #client = OpenAI()\n",
    "    #cleaned_text = clean_text_with_gpt4(docs)\n",
    "    #with open(file_name, 'w', encoding='utf-8') as file:\n",
    "    #    for text in cleaned_texts:\n",
    "    #        file.write(text + \"\\n\\n\")  # Adding two newlines as a separator between texts\n",
    "    #print(f\"Cleaned texts saved to {file_name}\")\n",
    "    cleaned_text = docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "78fd8749-3c13-4979-bf3a-ba71971323ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from langchain_community.vectorstores import Chroma\n",
    "\n",
    "text_splitter2 = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)\n",
    "splits = text_splitter2.split_documents(cleaned_text)\n",
    "vectorstore = Chroma.from_documents(documents=splits, embedding=OpenAIEmbeddings())\n",
    "CDC_retriever = vectorstore.as_retriever()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b73af828-adc6-4250-95ef-303534a6f449",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "100%|██████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 46.05it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n"
     ]
    }
   ],
   "source": [
    "import zipfile\n",
    "from langchain_community.document_loaders import DirectoryLoader\n",
    "\n",
    "loader2 = DirectoryLoader('./EpiHiper-Schema-master', glob=\"**/*.py\", show_progress=True)\n",
    "abm_codes = loader2.load()\n",
    "\n",
    "#code_splits = split_by_length_with_overlap(abm_codes)\n",
    "vectorstore = Chroma.from_documents(documents=abm_codes, embedding=OpenAIEmbeddings())\n",
    "ABM_retriever = vectorstore.as_retriever()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e223974-324a-4aac-8515-c438860c80a3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bf065cfd-7486-43dd-8f3a-c7ac2733b0cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_openai import ChatOpenAI\n",
    "from crewai import Agent, Task, Crew, Process\n",
    "import textwrap\n",
    "\n",
    "# llm = ChatOpenAI(model=\"gpt-3.5-turbo\", temperature=.2)\n",
    "llm = ChatOpenAI(model=\"gpt-4-1106-preview\", temperature=0.0) #\"gpt-4-1106-preview\", \"gpt-4-0314\"\n",
    "\n",
    "query_planner = Agent(\n",
    "    role=\"Simulation Planner\",\n",
    "    goal=\"Plan the steps needed to parameterize an agent-based simulation from existing knowledge\",\n",
    "    backstory=textwrap.dedent(\"\"\"\n",
    "        You are an expert at identifying modeling parameters from code base that implements \n",
    "        an agent-based model and listing the model choices, parameters, and json config files \n",
    "        whose values need to be determined. You will break down each model choice, parameter, \n",
    "        and json config file into sub-questions such that the answer to each sub-question will \n",
    "        inform the value to be used in the agent-based simulation.\n",
    "        Accept the user-question and determine if it requires sub-questions to either the\n",
    "        CDC website which provides an official source of recent infectious disease outbreaks\n",
    "        or Wikipedia for information about a geographical location, country, infectious agent\n",
    "        characteristics, transmission dynamics, infection states, or other epidemiological\n",
    "        modeling efforts.\n",
    "        Your final answer MUST be a description of sub-questions that explain the best model\n",
    "        choices, model parameters, and config files for an agent-based modeling code base.\n",
    "    \"\"\"),\n",
    "    verbose=True,\n",
    "    allow_delegation=False,\n",
    "    tools=[],  ###\n",
    "    llm=llm,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "905d12a7-d564-42c6-91a7-5f2cb284319d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#from langchain import hub\n",
    "#prompt = hub.pull(\"hwchase17/self-ask-with-search\")\n",
    "#print(f'{prompt.format(agent_scratchpad = \"AGENTSCRATCHPAD\", input = \"INPUT\")}')\n",
    "\n",
    "from langchain.tools.retriever import create_retriever_tool\n",
    "from langchain.tools import DuckDuckGoSearchRun\n",
    "\n",
    "CDC_retriever_tool = create_retriever_tool(\n",
    "    CDC_retriever,\n",
    "    \"CDC_retriever_tool\",\n",
    "    \"\"\"As an AI assistant you provide answers based on the given context, ensuring accuracy and briefness. \n",
    "\n",
    "        You always follow these guidelines:\n",
    "\n",
    "        -If the answer isn't available within the context, state that fact\n",
    "        -Otherwise, answer to your best capability, refering to source of documents provided\n",
    "        -Only use examples if explicitly requested\n",
    "        -Do not introduce examples outside of the context\n",
    "        -Do not answer if context is absent\n",
    "        -Limit responses to three or four sentences for clarity and conciseness\n",
    "        \n",
    "        Search for data related to outbreaks. For questions about outbreaks, use this tool to return \n",
    "        relevant data for answering questions about outbreaks\"\"\",\n",
    ")\n",
    "\n",
    "ABM_retriever_tool = create_retriever_tool(\n",
    "    ABM_retriever,\n",
    "    \"ABM_retriever_tool\",\n",
    "    \"\"\"As an AI assistant you provide answers based on the given context, ensuring accuracy and briefness. \n",
    "\n",
    "        You always follow these guidelines:\n",
    "\n",
    "        -If the answer isn't available within the context, state that fact\n",
    "        -Otherwise, answer to your best capability, refering to source of documents provided\n",
    "        -Only use examples if explicitly requested\n",
    "        -Do not introduce examples outside of the context\n",
    "        -Do not answer if context is absent\n",
    "        -Limit responses to three or four sentences for clarity and conciseness\n",
    "        \n",
    "        Search for data related to outbreaks. For questions about outbreaks, use this tool to return \n",
    "        relevant data for answering questions about outbreaks\"\"\",\n",
    ")\n",
    "\n",
    "web_search = DuckDuckGoSearchRun()\n",
    "\n",
    "query_executor = Agent(\n",
    "    role=\"\"\"\n",
    "        Agent Role: Information Searcher\n",
    "\n",
    "        Primary Objectives:\n",
    "        1. Utilize the CDC_retriever_tool to gather current outbreak information. This includes statistics, affected regions, and latest guidelines related to the outbreak.\n",
    "\n",
    "        2. Employ the ABM_retriever_tool to access the codebase for agent-based modeling simulations. Extract relevant parameters and settings that are crucial for understanding the dynamics of the disease spread in the simulations.\n",
    "\n",
    "        3. Conduct thorough internet searches using the web_search tool. Focus on disease-specific information such as modes of transmission, vectors involved, the role of asymptomatic infectious carriers, and insights from past modeling efforts.\n",
    "\n",
    "        Key Responsibilities:\n",
    "        - Ensure accurate and up-to-date information is retrieved from each tool.\n",
    "        - Synthesize information from diverse sources to provide a comprehensive understanding of the disease and its impact.\n",
    "        - Adhere to the principles of clarity and conciseness in reporting findings.\n",
    "        \"\"\",\n",
    "    goal=\"Information Searcher\",\n",
    "    backstory=textwrap.dedent(\"\"\"\n",
    "        Accept list of sub-questions from the query_planner agent and perform\n",
    "        the necessary searches to answer the questions.\n",
    "        Perform the tasks in the order given and report the result out.\n",
    "        Your final answer MUST be a correct response to the original user-query.\n",
    "    \"\"\"),\n",
    "    verbose=True,\n",
    "    llm=llm,\n",
    "    # tools=[SqlTools.do_sql_query, RagTools.do_rag_query],\n",
    "    tools=[CDC_retriever_tool, ABM_retriever_tool, web_search],\n",
    "    allow_delegation=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcba109b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working Agent: Simulation Planner\n",
      "Starting Task: \n",
      "Your task is to plan out the necessary steps in order to fulfull the user task and \n",
      "create smaller prompts for other co-workers to follow. The result of this task should\n",
      "be an organized list of questions to be answered in order to fulfill the user request.\n",
      "    Model the current flavivirus outbreak using an agent based model\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new CrewAgentExecutor chain...\u001b[0m\n",
      "\u001b[32;1m\u001b[1;3m```\n",
      "Thought: Do I need to use a tool? No\n",
      "Final Answer: To model the current flavivirus outbreak using an agent-based model, we need to gather specific information to accurately parameterize the simulation. Here is a breakdown of the model choices, parameters, and config files that need to be determined, along with the sub-questions that will inform their values:\n",
      "\n",
      "1. Model Choice: Type of Agent-Based Model\n",
      "   - What is the scale of the outbreak (local, regional, national, international)?\n",
      "   - What is the primary mode of transmission for the flavivirus in question?\n",
      "   - What are the key demographics that need to be represented in the model?\n",
      "\n",
      "2. Parameters: Epidemiological Characteristics\n",
      "   - What is the basic reproduction number (R0) of the current flavivirus outbreak?\n",
      "   - What is the incubation period of the flavivirus?\n",
      "   - What are the rates of transmission, recovery, and mortality associated with the flavivirus?\n",
      "   - What is the duration of infectiousness for individuals infected with the flavivirus?\n",
      "\n",
      "3. Parameters: Population Dynamics\n",
      "   - What is the population size and density in the affected area?\n",
      "   - What are the age and social structure distributions of the population?\n",
      "   - What are the mobility patterns or travel behaviors within the population?\n",
      "\n",
      "4. Parameters: Intervention Strategies\n",
      "   - What types of interventions (vaccination, vector control, quarantine, etc.) are being implemented?\n",
      "   - What is the coverage and efficacy of these interventions?\n",
      "\n",
      "5. JSON Config Files: Initialization and Runtime Settings\n",
      "   - What initial conditions need to be set for the simulation (e.g., initial number of infected individuals)?\n",
      "   - What is the time step for the simulation?\n",
      "   - What is the total duration for which the simulation will run?\n",
      "   - What are the output data requirements (e.g., number of infected individuals over time, peak infection rate)?\n",
      "\n",
      "6. JSON Config Files: Environmental and Seasonal Factors\n",
      "   - What are the environmental factors that could influence the transmission (e.g., temperature, rainfall)?\n",
      "   - Is there any seasonality in the transmission of the flavivirus?\n",
      "\n",
      "7. JSON Config Files: Data Sources and Calibration\n",
      "   - What are the sources of data for calibrating the model (e.g., CDC, WHO, peer-reviewed literature)?\n",
      "   - How will the model be validated against real-world data?\n",
      "\n",
      "By answering these sub-questions, we can determine the best model choices, parameters, and config file values for simulating the current flavivirus outbreak using an agent-based model. Each sub-question may require further research or data collection, potentially involving consultation with the CDC website for recent outbreak data or Wikipedia for information on geographical locations and other epidemiological modeling efforts.\n",
      "```\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "[Simulation Planner] Task output: To model the current flavivirus outbreak using an agent-based model, we need to gather specific information to accurately parameterize the simulation. Here is a breakdown of the model choices, parameters, and config files that need to be determined, along with the sub-questions that will inform their values:\n",
      "\n",
      "1. Model Choice: Type of Agent-Based Model\n",
      "   - What is the scale of the outbreak (local, regional, national, international)?\n",
      "   - What is the primary mode of transmission for the flavivirus in question?\n",
      "   - What are the key demographics that need to be represented in the model?\n",
      "\n",
      "2. Parameters: Epidemiological Characteristics\n",
      "   - What is the basic reproduction number (R0) of the current flavivirus outbreak?\n",
      "   - What is the incubation period of the flavivirus?\n",
      "   - What are the rates of transmission, recovery, and mortality associated with the flavivirus?\n",
      "   - What is the duration of infectiousness for individuals infected with the flavivirus?\n",
      "\n",
      "3. Parameters: Population Dynamics\n",
      "   - What is the population size and density in the affected area?\n",
      "   - What are the age and social structure distributions of the population?\n",
      "   - What are the mobility patterns or travel behaviors within the population?\n",
      "\n",
      "4. Parameters: Intervention Strategies\n",
      "   - What types of interventions (vaccination, vector control, quarantine, etc.) are being implemented?\n",
      "   - What is the coverage and efficacy of these interventions?\n",
      "\n",
      "5. JSON Config Files: Initialization and Runtime Settings\n",
      "   - What initial conditions need to be set for the simulation (e.g., initial number of infected individuals)?\n",
      "   - What is the time step for the simulation?\n",
      "   - What is the total duration for which the simulation will run?\n",
      "   - What are the output data requirements (e.g., number of infected individuals over time, peak infection rate)?\n",
      "\n",
      "6. JSON Config Files: Environmental and Seasonal Factors\n",
      "   - What are the environmental factors that could influence the transmission (e.g., temperature, rainfall)?\n",
      "   - Is there any seasonality in the transmission of the flavivirus?\n",
      "\n",
      "7. JSON Config Files: Data Sources and Calibration\n",
      "   - What are the sources of data for calibrating the model (e.g., CDC, WHO, peer-reviewed literature)?\n",
      "   - How will the model be validated against real-world data?\n",
      "\n",
      "By answering these sub-questions, we can determine the best model choices, parameters, and config file values for simulating the current flavivirus outbreak using an agent-based model. Each sub-question may require further research or data collection, potentially involving consultation with the CDC website for recent outbreak data or Wikipedia for information on geographical locations and other epidemiological modeling efforts.\n",
      "```\n",
      "\n",
      "\n",
      "Working Agent: \n",
      "        Agent Role: Information Searcher\n",
      "\n",
      "        Primary Objectives:\n",
      "        1. Utilize the CDC_retriever_tool to gather current outbreak information. This includes statistics, affected regions, and latest guidelines related to the outbreak.\n",
      "\n",
      "        2. Employ the ABM_retriever_tool to access the codebase for agent-based modeling simulations. Extract relevant parameters and settings that are crucial for understanding the dynamics of the disease spread in the simulations.\n",
      "\n",
      "        3. Conduct thorough internet searches using the web_search tool. Focus on disease-specific information such as modes of transmission, vectors involved, the role of asymptomatic infectious carriers, and insights from past modeling efforts.\n",
      "\n",
      "        Key Responsibilities:\n",
      "        - Ensure accurate and up-to-date information is retrieved from each tool.\n",
      "        - Synthesize information from diverse sources to provide a comprehensive understanding of the disease and its impact.\n",
      "        - Adhere to the principles of clarity and conciseness in reporting findings.\n",
      "        \n",
      "Starting Task: \n",
      "You will recieve a set of queries from the previous task. Your task is to update these\n",
      "queries based on the latest information from the CDC database using the CDC_retriever_tool\n",
      "tool and from a web search using the web_search tool in order to either make the queries \n",
      "more specific or to provide the answers so they can be passed on to the next query.\n",
      "The information gathered should include key statistics such as number of cases, affected \n",
      "regions, and any new guidelines issued. \n",
      "Prepare the necessary sub-queries or search queries to handle this user-query:\n",
      "    Model the current flavivirus outbreak using an agent based model\n",
      "You should use the CDC_retriever_tool and web_search to access the relevant data.\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new CrewAgentExecutor chain...\u001b[0m\n",
      "\u001b[32;1m\u001b[1;3mThought: Do I need to use a tool? Yes\n",
      "Action: CDC_retriever_tool\n",
      "Action Input: current flavivirus outbreak statistics affected regions guidelines\u001b[0m\u001b[36;1m\u001b[1;3m[Document(page_content='Crimean-Congo Hemorrhagic Fever, Spain, 2013–2021\\n\\n\\n\\n\\n\\n\\n\\r\\n        [PDF - 2.14 MB - 8 pages]\\r\\n    \\n\\n\\nH. Lorenzo Juanes et al.\\n View Abstract\\n Cite This Article', metadata={'description': 'Emerging Infectious Diseases is a peer-reviewed, monthly journal published by the Centers for Disease Control and Prevention (CDC). It offers global health professionals the latest scientific information on emerging infectious diseases and trends. Articles provide the most up-to-date information on infectious diseases and their effects on global health.', 'language': 'en-us', 'source': 'https://www.cdc.gov/eid/', 'title': 'Emerging Infectious Diseases - CDC'}), Document(page_content='Crimean-Congo Hemorrhagic Fever, Spain, 2013–2021\\n\\n\\n\\n\\n\\n\\n\\r\\n        [PDF - 2.14 MB - 8 pages]\\r\\n    \\n\\n\\nH. Lorenzo Juanes et al.\\n View Abstract\\n Cite This Article', metadata={'description': 'Emerging Infectious Diseases is a peer-reviewed, monthly journal published by the Centers for Disease Control and Prevention (CDC). It offers global health professionals the latest scientific information on emerging infectious diseases and trends. Articles provide the most up-to-date information on infectious diseases and their effects on global health.', 'language': 'en-us', 'source': 'https://www.cdc.gov/eid/', 'title': 'Emerging Infectious Diseases - CDC'}), Document(page_content='Introduction and Spread of Dengue Virus 3, Florida, USA, May 2022–April 2023\\r\\n    \\nF. K. Jones et al.\\n\\n\\n\\r\\n        Tick-borne Encephalitis, Lombardy, Italy\\r\\n    \\nA. Gaffuri et al.\\n\\n\\n\\r\\n        Power Law for Estimating Underdetection of Foodborne Disease Outbreaks, United States\\r\\n    \\nL. Ford et al.\\n\\n\\n\\r\\n        Confirmed Autochthonous Case of Human Alveolar Echinococcosis, Italy, 2023\\r\\n    \\nF. Tamarozzi et al.\\n\\n\\n\\nResearch Letters\\n\\n\\r\\n        Lymphocytic Choriomeningitis Virus Lineage V in Wood Mice, Germany\\r\\n    \\nC. Mehl et al.\\n\\n\\n\\nPhylogenomics of Dengue Virus Isolates Causing Dengue Outbreak, São Tomé and Príncipe, 2022\\n\\nL. Lázaro et al.\\n View Abstract\\n\\n\\n\\nWe determined that the dengue outbreak in São Tomé and Príncipe during 2022 was caused by dengue virus serotype 3 genotype III. Phylogenomic analyses showed that the outbreak strain was closely related to the newly identified GIII-American-II lineage and that the virus probably was introduced from the Americas.', metadata={'description': 'Emerging Infectious Diseases is a peer-reviewed, monthly journal published by the Centers for Disease Control and Prevention (CDC). It offers global health professionals the latest scientific information on emerging infectious diseases and trends. Articles provide the most up-to-date information on infectious diseases and their effects on global health.', 'language': 'en-us', 'source': 'https://www.cdc.gov/eid/', 'title': 'Emerging Infectious Diseases - CDC'}), Document(page_content='Introduction and Spread of Dengue Virus 3, Florida, USA, May 2022–April 2023\\r\\n    \\nF. K. Jones et al.\\n\\n\\n\\r\\n        Tick-borne Encephalitis, Lombardy, Italy\\r\\n    \\nA. Gaffuri et al.\\n\\n\\n\\r\\n        Power Law for Estimating Underdetection of Foodborne Disease Outbreaks, United States\\r\\n    \\nL. Ford et al.\\n\\n\\n\\r\\n        Confirmed Autochthonous Case of Human Alveolar Echinococcosis, Italy, 2023\\r\\n    \\nF. Tamarozzi et al.\\n\\n\\n\\nResearch Letters\\n\\n\\r\\n        Lymphocytic Choriomeningitis Virus Lineage V in Wood Mice, Germany\\r\\n    \\nC. Mehl et al.\\n\\n\\n\\nPhylogenomics of Dengue Virus Isolates Causing Dengue Outbreak, São Tomé and Príncipe, 2022\\n\\nL. Lázaro et al.\\n View Abstract\\n\\n\\n\\nWe determined that the dengue outbreak in São Tomé and Príncipe during 2022 was caused by dengue virus serotype 3 genotype III. Phylogenomic analyses showed that the outbreak strain was closely related to the newly identified GIII-American-II lineage and that the virus probably was introduced from the Americas.', metadata={'description': 'Emerging Infectious Diseases is a peer-reviewed, monthly journal published by the Centers for Disease Control and Prevention (CDC). It offers global health professionals the latest scientific information on emerging infectious diseases and trends. Articles provide the most up-to-date information on infectious diseases and their effects on global health.', 'language': 'en-us', 'source': 'https://www.cdc.gov/eid/', 'title': 'Emerging Infectious Diseases - CDC'})]\u001b[0m\u001b[32;1m\u001b[1;3mDo I need to use a tool? No\n",
      "Final Answer: The CDC database has provided information on various flavivirus outbreaks, including Dengue Virus 3 in Florida, USA, from May 2022 to April 2023, and an outbreak in São Tomé and Príncipe during 2022 caused by Dengue Virus serotype 3 genotype III. However, there are no specific statistics, affected regions, or new guidelines mentioned in the documents retrieved. Further searches may be necessary to gather the required detailed information for modeling the current flavivirus outbreak.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "[\n",
      "        Agent Role: Information Searcher\n",
      "\n",
      "        Primary Objectives:\n",
      "        1. Utilize the CDC_retriever_tool to gather current outbreak information. This includes statistics, affected regions, and latest guidelines related to the outbreak.\n",
      "\n",
      "        2. Employ the ABM_retriever_tool to access the codebase for agent-based modeling simulations. Extract relevant parameters and settings that are crucial for understanding the dynamics of the disease spread in the simulations.\n",
      "\n",
      "        3. Conduct thorough internet searches using the web_search tool. Focus on disease-specific information such as modes of transmission, vectors involved, the role of asymptomatic infectious carriers, and insights from past modeling efforts.\n",
      "\n",
      "        Key Responsibilities:\n",
      "        - Ensure accurate and up-to-date information is retrieved from each tool.\n",
      "        - Synthesize information from diverse sources to provide a comprehensive understanding of the disease and its impact.\n",
      "        - Adhere to the principles of clarity and conciseness in reporting findings.\n",
      "        ] Task output: The CDC database has provided information on various flavivirus outbreaks, including Dengue Virus 3 in Florida, USA, from May 2022 to April 2023, and an outbreak in São Tomé and Príncipe during 2022 caused by Dengue Virus serotype 3 genotype III. However, there are no specific statistics, affected regions, or new guidelines mentioned in the documents retrieved. Further searches may be necessary to gather the required detailed information for modeling the current flavivirus outbreak.\n",
      "\n",
      "\n",
      "Working Agent: \n",
      "        Agent Role: Information Searcher\n",
      "\n",
      "        Primary Objectives:\n",
      "        1. Utilize the CDC_retriever_tool to gather current outbreak information. This includes statistics, affected regions, and latest guidelines related to the outbreak.\n",
      "\n",
      "        2. Employ the ABM_retriever_tool to access the codebase for agent-based modeling simulations. Extract relevant parameters and settings that are crucial for understanding the dynamics of the disease spread in the simulations.\n",
      "\n",
      "        3. Conduct thorough internet searches using the web_search tool. Focus on disease-specific information such as modes of transmission, vectors involved, the role of asymptomatic infectious carriers, and insights from past modeling efforts.\n",
      "\n",
      "        Key Responsibilities:\n",
      "        - Ensure accurate and up-to-date information is retrieved from each tool.\n",
      "        - Synthesize information from diverse sources to provide a comprehensive understanding of the disease and its impact.\n",
      "        - Adhere to the principles of clarity and conciseness in reporting findings.\n",
      "        \n",
      "Starting Task: \n",
      "You will receive a set of queries from the previous task. \n",
      "Use the ABM_retriever_tool to query the agent-based modeling codebase for parameters \n",
      "and settings related to disease spread simulation. Report the parameters in the format\n",
      "required for input into the code base that is accessible from the ABM_retriever_tool.\n",
      "Your final answer must be a list of parameters and values or information that can be\n",
      "used to estimate those values in the appropriate format:\n",
      "    Model the current flavivirus outbreak using an agent based model\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new CrewAgentExecutor chain...\u001b[0m\n",
      "\u001b[32;1m\u001b[1;3mThought: Do I need to use a tool? Yes\n",
      "Action: ABM_retriever_tool\n",
      "Action Input: Model the current flavivirus outbreak using an agent based model\u001b[0m\u001b[33;1m\u001b[1;3m[Document(page_content='To support the ongoing management of viral respiratory diseases while transitioning out of the acute phase of the COVID-19 pandemic, many countries are moving toward an integrated model of surveillance for SARS-CoV-2, influenza virus, and other respiratory pathogens. Although many surveillance approaches catalyzed by the COVID-19 pandemic provide novel epidemiologic insight, continuing them as implemented during the pandemic is unlikely to be feasible for nonemergency surveillance, and many have already been scaled back. Furthermore, given anticipated cocirculation of SARS-CoV-2 and influenza virus, surveillance activities in place before the pandemic require review and adjustment to ensure their ongoing value for public health. In this report, we highlight key challenges for the development of integrated models of surveillance. We discuss the relative strengths and limitations of different surveillance practices and studies as well as their contribution to epidemiologic assessment,', metadata={'description': 'Emerging Infectious Diseases is a peer-reviewed, monthly journal published by the Centers for Disease Control and Prevention (CDC). It offers global health professionals the latest scientific information on emerging infectious diseases and trends. Articles provide the most up-to-date information on infectious diseases and their effects on global health.', 'language': 'en-us', 'source': 'https://www.cdc.gov/eid/', 'title': 'Emerging Infectious Diseases - CDC'}), Document(page_content='To support the ongoing management of viral respiratory diseases while transitioning out of the acute phase of the COVID-19 pandemic, many countries are moving toward an integrated model of surveillance for SARS-CoV-2, influenza virus, and other respiratory pathogens. Although many surveillance approaches catalyzed by the COVID-19 pandemic provide novel epidemiologic insight, continuing them as implemented during the pandemic is unlikely to be feasible for nonemergency surveillance, and many have already been scaled back. Furthermore, given anticipated cocirculation of SARS-CoV-2 and influenza virus, surveillance activities in place before the pandemic require review and adjustment to ensure their ongoing value for public health. In this report, we highlight key challenges for the development of integrated models of surveillance. We discuss the relative strengths and limitations of different surveillance practices and studies as well as their contribution to epidemiologic assessment,', metadata={'description': 'Emerging Infectious Diseases is a peer-reviewed, monthly journal published by the Centers for Disease Control and Prevention (CDC). It offers global health professionals the latest scientific information on emerging infectious diseases and trends. Articles provide the most up-to-date information on infectious diseases and their effects on global health.', 'language': 'en-us', 'source': 'https://www.cdc.gov/eid/', 'title': 'Emerging Infectious Diseases - CDC'}), Document(page_content='Introduction and Spread of Dengue Virus 3, Florida, USA, May 2022–April 2023\\r\\n    \\nF. K. Jones et al.\\n\\n\\n\\r\\n        Tick-borne Encephalitis, Lombardy, Italy\\r\\n    \\nA. Gaffuri et al.\\n\\n\\n\\r\\n        Power Law for Estimating Underdetection of Foodborne Disease Outbreaks, United States\\r\\n    \\nL. Ford et al.\\n\\n\\n\\r\\n        Confirmed Autochthonous Case of Human Alveolar Echinococcosis, Italy, 2023\\r\\n    \\nF. Tamarozzi et al.\\n\\n\\n\\nResearch Letters\\n\\n\\r\\n        Lymphocytic Choriomeningitis Virus Lineage V in Wood Mice, Germany\\r\\n    \\nC. Mehl et al.\\n\\n\\n\\nPhylogenomics of Dengue Virus Isolates Causing Dengue Outbreak, São Tomé and Príncipe, 2022\\n\\nL. Lázaro et al.\\n View Abstract\\n\\n\\n\\nWe determined that the dengue outbreak in São Tomé and Príncipe during 2022 was caused by dengue virus serotype 3 genotype III. Phylogenomic analyses showed that the outbreak strain was closely related to the newly identified GIII-American-II lineage and that the virus probably was introduced from the Americas.', metadata={'description': 'Emerging Infectious Diseases is a peer-reviewed, monthly journal published by the Centers for Disease Control and Prevention (CDC). It offers global health professionals the latest scientific information on emerging infectious diseases and trends. Articles provide the most up-to-date information on infectious diseases and their effects on global health.', 'language': 'en-us', 'source': 'https://www.cdc.gov/eid/', 'title': 'Emerging Infectious Diseases - CDC'}), Document(page_content='Introduction and Spread of Dengue Virus 3, Florida, USA, May 2022–April 2023\\r\\n    \\nF. K. Jones et al.\\n\\n\\n\\r\\n        Tick-borne Encephalitis, Lombardy, Italy\\r\\n    \\nA. Gaffuri et al.\\n\\n\\n\\r\\n        Power Law for Estimating Underdetection of Foodborne Disease Outbreaks, United States\\r\\n    \\nL. Ford et al.\\n\\n\\n\\r\\n        Confirmed Autochthonous Case of Human Alveolar Echinococcosis, Italy, 2023\\r\\n    \\nF. Tamarozzi et al.\\n\\n\\n\\nResearch Letters\\n\\n\\r\\n        Lymphocytic Choriomeningitis Virus Lineage V in Wood Mice, Germany\\r\\n    \\nC. Mehl et al.\\n\\n\\n\\nPhylogenomics of Dengue Virus Isolates Causing Dengue Outbreak, São Tomé and Príncipe, 2022\\n\\nL. Lázaro et al.\\n View Abstract\\n\\n\\n\\nWe determined that the dengue outbreak in São Tomé and Príncipe during 2022 was caused by dengue virus serotype 3 genotype III. Phylogenomic analyses showed that the outbreak strain was closely related to the newly identified GIII-American-II lineage and that the virus probably was introduced from the Americas.', metadata={'description': 'Emerging Infectious Diseases is a peer-reviewed, monthly journal published by the Centers for Disease Control and Prevention (CDC). It offers global health professionals the latest scientific information on emerging infectious diseases and trends. Articles provide the most up-to-date information on infectious diseases and their effects on global health.', 'language': 'en-us', 'source': 'https://www.cdc.gov/eid/', 'title': 'Emerging Infectious Diseases - CDC'})]\u001b[0m\u001b[32;1m\u001b[1;3mDo I need to use a tool? No\n",
      "Final Answer: The ABM_retriever_tool did not provide specific parameters and settings related to disease spread simulation for the current flavivirus outbreak. The documents retrieved discuss the management of viral respiratory diseases and surveillance practices, as well as specific details about the dengue virus serotype 3 genotype III outbreak in São Tomé and Príncipe and its phylogenomic analysis. However, these documents do not contain the necessary information to extract simulation parameters for an agent-based model. Further information would be required to model the current flavivirus outbreak accurately.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "[\n",
      "        Agent Role: Information Searcher\n",
      "\n",
      "        Primary Objectives:\n",
      "        1. Utilize the CDC_retriever_tool to gather current outbreak information. This includes statistics, affected regions, and latest guidelines related to the outbreak.\n",
      "\n",
      "        2. Employ the ABM_retriever_tool to access the codebase for agent-based modeling simulations. Extract relevant parameters and settings that are crucial for understanding the dynamics of the disease spread in the simulations.\n",
      "\n",
      "        3. Conduct thorough internet searches using the web_search tool. Focus on disease-specific information such as modes of transmission, vectors involved, the role of asymptomatic infectious carriers, and insights from past modeling efforts.\n",
      "\n",
      "        Key Responsibilities:\n",
      "        - Ensure accurate and up-to-date information is retrieved from each tool.\n",
      "        - Synthesize information from diverse sources to provide a comprehensive understanding of the disease and its impact.\n",
      "        - Adhere to the principles of clarity and conciseness in reporting findings.\n",
      "        ] Task output: The ABM_retriever_tool did not provide specific parameters and settings related to disease spread simulation for the current flavivirus outbreak. The documents retrieved discuss the management of viral respiratory diseases and surveillance practices, as well as specific details about the dengue virus serotype 3 genotype III outbreak in São Tomé and Príncipe and its phylogenomic analysis. However, these documents do not contain the necessary information to extract simulation parameters for an agent-based model. Further information would be required to model the current flavivirus outbreak accurately.\n",
      "\n",
      "\n",
      "Working Agent: \n",
      "        Agent Role: Information Searcher\n",
      "\n",
      "        Primary Objectives:\n",
      "        1. Utilize the CDC_retriever_tool to gather current outbreak information. This includes statistics, affected regions, and latest guidelines related to the outbreak.\n",
      "\n",
      "        2. Employ the ABM_retriever_tool to access the codebase for agent-based modeling simulations. Extract relevant parameters and settings that are crucial for understanding the dynamics of the disease spread in the simulations.\n",
      "\n",
      "        3. Conduct thorough internet searches using the web_search tool. Focus on disease-specific information such as modes of transmission, vectors involved, the role of asymptomatic infectious carriers, and insights from past modeling efforts.\n",
      "\n",
      "        Key Responsibilities:\n",
      "        - Ensure accurate and up-to-date information is retrieved from each tool.\n",
      "        - Synthesize information from diverse sources to provide a comprehensive understanding of the disease and its impact.\n",
      "        - Adhere to the principles of clarity and conciseness in reporting findings.\n",
      "        \n",
      "Starting Task: \n",
      "You will recieve a set of information from the previous task. Based on the information,\n",
      "your task is to populate the parameter keys and values into the final input format\n",
      "required for running the agent-based model.\n",
      "Your findings should directly support the original user-query:\n",
      "    Model the current flavivirus outbreak using an agent based model\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new CrewAgentExecutor chain...\u001b[0m\n",
      "\u001b[32;1m\u001b[1;3mThought: Do I need to use a tool? Yes\n",
      "Action: duckduckgo_search\n",
      "Action Input: current flavivirus outbreak statistics transmission mode affected regions guidelines\u001b[0m\u001b[38;5;200m\u001b[1;3mYellow fever virus is an RNA virus that belongs to the genus Flavivirus. It is related to West Nile, St. Louis encephalitis, and Japanese encephalitis viruses. Yellow fever virus is transmitted to people primarily through the bite of infected Aedes or Haemagogus species mosquitoes. Mosquitoes acquire the virus by feeding on infected primates ... Wastewater viral activity levels, an important tool used to detect increases of COVID-19 transmission in the community, are currently high and increasing in all regions. As of December 25, 2023, 66% of wastewater samples had JN.1 as the dominant variant, up from 58% the previous week. JN1.1 is not only rising in the United States, but globally ... The extensive global burden is predominantly caused by six flaviviruses: dengue, Zika, West Nile, yellow fever, Japanese encephalitis and tick-borne encephalitis. Several vaccines have been developed, and many more are currently being tested in clinical trials. Dengue is the most rapidly spreading mosquito-borne viral disease globally, with outbreaks becoming more frequent, more deadly and more widespread across diverse regions of the world. In 2017 alone, there were an estimated 100 million cases of Dengue infection across the world, 40,000 of which were fatal. The flavivirus genus contains several clinically important pathogens that account for tremendous global suffering. Primarily transmitted by mosquitos or ticks, these viruses can cause severe and potentially fatal diseases ranging from hemorrhagic fevers to encephalitis. The extensive global burden is predominantly caused by six flaviviruses ...\u001b[0m"
     ]
    }
   ],
   "source": [
    "user_query = \"Model the current flavivirus outbreak using an agent based model\"\n",
    "\n",
    "task0 = Task(\n",
    "    description=textwrap.dedent(f\"\"\"\n",
    "        Your task is to plan out the necessary steps in order to fulfull the user task and \n",
    "        create smaller prompts for other co-workers to follow. The result of this task should\n",
    "        be an organized list of questions to be answered in order to fulfill the user request.\n",
    "            {user_query}\n",
    "    \"\"\"),\n",
    "    agent=query_planner\n",
    ")\n",
    "\n",
    "task1 = Task(\n",
    "    description=textwrap.dedent(f\"\"\"\n",
    "        You will recieve a set of queries from the previous task. Your task is to update these\n",
    "        queries based on the latest information from the CDC database using the CDC_retriever_tool\n",
    "        tool and from a web search using the web_search tool in order to either make the queries \n",
    "        more specific or to provide the answers so they can be passed on to the next query.\n",
    "        The information gathered should include key statistics such as number of cases, affected \n",
    "        regions, and any new guidelines issued. \n",
    "        Prepare the necessary sub-queries or search queries to handle this user-query:\n",
    "            {user_query}\n",
    "        You should use the CDC_retriever_tool and web_search to access the relevant data.\n",
    "    \"\"\"),\n",
    "    agent=query_executor\n",
    ")\n",
    "\n",
    "task2 = Task(\n",
    "    description=textwrap.dedent(f\"\"\"\n",
    "        You will receive a set of queries from the previous task. \n",
    "        Use the ABM_retriever_tool to query the agent-based modeling codebase for parameters \n",
    "        and settings related to disease spread simulation. Report the parameters in the format\n",
    "        required for input into the code base that is accessible from the ABM_retriever_tool.\n",
    "        Your final answer must be a list of parameters and values or information that can be\n",
    "        used to estimate those values in the appropriate format:\n",
    "            {user_query}\n",
    "    \"\"\"),\n",
    "    agent=query_executor\n",
    ")\n",
    "\n",
    "task3 = Task(\n",
    "    description=textwrap.dedent(f\"\"\"\n",
    "        You will recieve a set of information from the previous task. Based on the information,\n",
    "        your task is to populate the parameter keys and values into the final input format\n",
    "        required for running the agent-based model.\n",
    "        Your findings should directly support the original user-query:\n",
    "            {user_query}\n",
    "    \"\"\"),\n",
    "    agent=query_executor\n",
    ")\n",
    "\n",
    "crew = Crew(\n",
    "    agents=[query_planner, query_executor],\n",
    "    tasks=[task0, task1, task2, task3],\n",
    "    verbose=2,  # print what tasks are being worked on, can set it to 1 or 2\n",
    "    process=Process.sequential,\n",
    ")\n",
    "\n",
    "result = crew.kickoff()\n",
    "\n",
    "print(\"######################\")\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "11a565bf-bad9-41c0-b372-0d579a21fdaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tool(name='Intermediate Answer', description=\"As an AI assistant you provide answers based on the given context, ensuring accuracy and briefness. \\n\\n        You always follow these guidelines:\\n\\n        -If the answer isn't available within the context, state that fact\\n        -Otherwise, answer to your best capability, refering to source of documents provided\\n        -Only use examples if explicitly requested\\n        -Do not introduce examples outside of the context\\n        -Do not answer if context is absent\\n        -Limit responses to three or four sentences for clarity and conciseness\\n        \\n        Search for data related to outbreaks. For questions about outbreaks, use this tool to return \\n        relevant data for answering questions about outbreaks\", args_schema=<class 'langchain.tools.retriever.RetrieverInput'>, func=<bound method BaseRetriever.get_relevant_documents of VectorStoreRetriever(tags=['Chroma', 'OpenAIEmbeddings'], vectorstore=<langchain_community.vectorstores.chroma.Chroma object at 0x11e3cf450>)>, coroutine=<bound method BaseRetriever.aget_relevant_documents of VectorStoreRetriever(tags=['Chroma', 'OpenAIEmbeddings'], vectorstore=<langchain_community.vectorstores.chroma.Chroma object at 0x11e3cf450>)>)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDC_retriever_tool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "2ebd8839-932d-4c0e-a114-7243dd22f2da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DuckDuckGoSearchRun()"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "web_search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66a37703-f14f-4e0b-8c65-ce1c74135186",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
